### ISSUE TITLE
OSD in fullscreen mode stops  HTML5 background video

### ISSUE THREAD

**Issue by comrade2sun** [openseadragon/openseadragon#1050]
#### Some experimental stuff.
As I noticed in some other cases before, OSD stops HTML5 videos from playing when you go fullscreen. 
Just check this web page for a better understanding what I mean:

<a href="http://homepage.o2mail.de/nixneues/OSD/test/dd_full.html">HTML5 Video - Test</a>

First hit ENTER or the F-key to toggle fullscreen _(custom toggle fullscreen script, usally I provide a button)_.
The result: No effect to the video!

Next do the same with the default OSD button.
The result: The video stops! _(reload for start playing again)_

I do get this effects in Firefox 49, Google Chrome 54.0, Microsoft Edge 38, Internet Explorer 11. This means, it's not some browser related "wiggle di waggle" and on the other hand, I don't get what's wrong with <a href="https://github.com/openseadragon/openseadragon/blob/master/src/fullscreen.js">openseadragon/fullscreen.js </a>

In this mentioned test scenario I could simply omit the default OSD script and live with my custom toggle fullscreen script. But there are some other imaginable scenarios and they are not that simple. For this reason I would prefer to find out what is causing this strange behavior.
Any ideas, suggestions and help welcome.

Thanks

**Comment by iangilman** [comment]
When OSD goes into fullscreen, it first engages its "full page" mode, which pulls all of the elements off of the body and hides them away, and attaches the OSD viewer directly to the body. All of this shuffling around probably interrupts the video. In fact, since the video isn't part of the viewer, it's one of the elements that gets hidden.

You don't really need any of that, so probably just going with your own full screen code is a fine choice (you'll just want to hook it up to the button as well).

**Comment by comrade2sun** [comment]
#### Moin Ian
For the above provided "kind of standalone" webpage a custom script would be sufficient. **But ...**
... the point is, OSD stops any other video on the webpage!

Let's take it one step further:

<a href="http://homepage.o2mail.de/nixneues/OSD/test/dd_test.html">HTML5 Video - Test II</a>

—  and for the sake of experiments a different setup:

- OSD with a background video
- Text animation with background video
- Youtube video

We got now 3 possibilities to go full screen:

- My custom  = button top right
- OSD  = button in the OSD-toolbar
- Youtube = button in the Yt-toolbar

####  Full screen mania here we go:
Press the button top right ➝ press the full screen button in the OSD toolbar = **_no reaction!_**
While still in full screen mode scroll past the animated text, start the youtube video ➝ use the youtube full screen button = **_youtube goes full screen!_**
Toggle the youtube video via its button = **_no effect on the other 2 background videos!_**
Now leave the full screen mode via the button top right while the youtube video is still playing.
Next press the fullscreen button in the OSD-toolbar ➝ **_OSD goes full screen and stops all other 3 videos!_** ... as you can see, when you leave full screen mode.

This test-scenario demonstrates 2 issues: _(assuming the visitors are the same full screen freaks as I am)_
While in full screen mode allready, the visitor is not able to go full screen with any OSD somewhere else on the webpage, (as youtube is doing without trouble).
OSD stops any other video on the webpage.

#### To be clear:
The internet is not going to break down in case I don't find any solution, hack or Plan B for this issues. **But ...**
... I would like to find out a little bit more, how  OSD "pulls elements off the body" and why videos stop playing.

Any hints welcome. Thanks

**Comment by iangilman** [comment]
Good point. Though really, for actual video you might be watching, it kind of seems like a good feature that it stops when it gets obscured. I realize that's not the case for ambient video that you just want to keep looping (though ideally that too would pause when it gets obscured to save on cycles and bandwidth, and then start up again when it's visible again).

At any rate, I agree OSD shouldn't be pausing the videos if we can avoid it. Time for a history lesson: 

Once upon a time, when OpenSeadragon was young, there was no fullscreen API for web browsers. The best you could do would be to expand to fill the web page. The same button we have now to go to fullscreen did "full page" at that time. The most reliable way to accomplish this is to remove everything else from the page (storing it offscreen until you were done with fullpage mode) and resize the viewer to fill the page.

Gradually browsers added fullscreen support, so the fullpage feature was extended to also go to fullscreen. For a while some browsers had fullscreen and others didn't, so it was important to keep this hybrid approach.

Nowadays fullscreen is pretty universal (except old versions of IE, and we do still support IE8 so we're not entirely out of the woods yet), so perhaps it's time to reevaluate our fullpage/fullscreen feature.

If, when we went to fullscreen, we didn't first do the fullpage thing, I believe we would get all of our fullscreen goodness without any of the videos stopping (since we're not mucking with the video's elements). Perhaps it's time to take that step.

We'd want to deprecate the fullpage functions properly, since they're still part of the API, and we'd want to still have the fullpage functionality as a fallback on old browsers, but for modern browsers we can skip the fullpage step.

What do you think? Would you be up for working on this?

**Comment by comrade2sun** [comment]
#### Hi Ian 
Thanks for your explanation.
What puzzled me most, was this "pulling of the other elements". Way back in my mind I knew,  with the new Fullscreen API there was this feature _"expanding any element into full screen"_, even when the body is in full screen. That's why this youtube video came as an example in to this test case.

Anyway I am already on "Plan B" and omit the OSD default full page feature.
Eh voilà, with a bit of searching I found this script:
> <a href="https://github.com/sindresorhus/screenfull.js">GitHub - sindresorhus/screenfull.js: Simple wrapper for cross-browser usage of the JavaScript Fullscreen API</a>

Right now <a href="http://homepage.o2mail.de/nixneues/OSD/test/dd_test.html">it works like a charm on firefox 49</a> and basically it should work on most modern browsers:
> <a href="http://caniuse.com/#search=full%20screen%20api">Can I use... Full Screen API?</a>

 I got some glitches on Chrome 54 and IE 11, but I think this is more related to some CSS issues and my fancy pille palle setup. _(And of course this ambient video stuff is a bit exaggerated  ... so better to hide it from IE & Edge)_  ;~) 

Speaking about OSD, I suggest you should use the new 
> <a href="https://fullscreen.spec.whatwg.org/">Fullscreen API Standard</a>

... but I am aware, this is just easy talk, because I can give a damn about IE, Safari and all the other *droids out there.

 Definitely I will spend some more time on this topic  and let you know if there is any other progress.

**Comment by iangilman** [comment]
As far as I know we do use the browser's standard fullscreen API, but we polyfill it with this file: 

https://github.com/openseadragon/openseadragon/blob/master/src/fullscreen.js

...resulting in this low level API:

http://openseadragon.github.io/docs/OpenSeadragon.html#.fullScreenApi

...which you can use instead of screenfull.js as your "Plan B" if you'd like.

I still think untangling our fullscreen feature from our fullpage feature is probably worth doing at some point...

**Comment by comrade2sun** [comment]
#### On y va ...
_"Le plus beau des droits de l'homme est le droit de pêche."_

I prefer to stick to the screenfull.js polyfill for certain reasons and because it offers me some more options for future experiments.
To be honest:
One of the main reasons is, I dont know how to write a line  of code that does the same as I am doing right now. The code sits there in "openseadragon/fullscreen.js", but how the hell can I shoot the fish in this barrel without a shotgun? ... means without any JavaScript skills?!

#### Let's go fly fishing ...
Instead of expanding the documents body with <a href="https://github.com/openseadragon/openseadragon/blob/master/src/viewer.js#L1155">$.requestFullScreen( document.body );</a> and pulling all the other stuff off the szene - like the default OSD fullscreen behavior is - I should be able to expand the OSD div-Element and leaving the other stuff underneath untouched.
This solves in my case the "fullscreen in fullscreen mode" and "ambient video pille palle" problem.
I know, rare use cases ...  "Confiteor […] <strike>mea culpa, mea maxima culpa</strike> …“ But ...
1 other thing comes in crafty: I don't depend on the .fullscreen CSS-Class anymore _(anyhow I hope so)_.

#### The <strike>trouts</strike> frogs I catched so far ...
<a href="http://homepage.o2mail.de/nixneues/OSD/test/dd_test.html">HTML5 Video - TEST № 3</a>

My test abilities are limited to Firefox 49 | Edge 38 | IE11 and Chrome 54. None of them in this menagerie has problems expanding the osd-div element. It becomes a little tricky when the osd-div is nested in an extra wrapper,  and the only one that is behaving come una donna capricciosa so far, is Chrome ... but that's another story.

 &iquest; 3 Questions in 1?
How can I achieve the same with OSD-fullscreen.js witout pulling other stuff off the document and not breaking the fallback for older browsers and mobiles  ... ?
_(1 more: How to present this options crystal clear to other potential "fish in a barrel" hunters like me?)_

Gone fishing ... a least for a while ♫

**Comment by iangilman** [comment]
Well, I think there's no problem in using screenfull.js if you prefer. I was just saying that OSD has a low level utility function like that, that the high level function (which does the pulling things off the scene) calls. To use the low level function would be something like:

```
OpenSeadragon.requestFullScreen(element);
```

...but sticking with screenfull.js is fine too.

As for a more comprehensive fix, I suppose we can wait for someone more confident in digging around and reorganizing the OSD code.

**Comment by comrade2sun** [comment]
Indeed, I suggest you shouldn't rely on my kind of "expertise"!  :innocent: 
Secondly I think this case was quite a bit exotic. There is no hurry to turn everything upside down and so far I found the solution.

Just for the reason of learning more, I will dig deeper in the viewer.js, because as more as I understand  the internal working of OSD, the  better I will handle any future experiments _(... and lesser the foolish questions)_.

Once again, this conversation was a great pleasure, because I made some more progress with OSD.

Thanks for your insightful replies

**Comment by iangilman** [comment]
Sounds good... I welcome your interest; the more people understand the internals, the better off we all are!

**Comment by R4wizard** [comment]
We had this same issue on our project. As a workaround we've simply done the following:

```js
let panZoomRef = document.querySelector(".pan-zoom-widget")
let sourceNode = panZoomRef.querySelector('div[title="Toggle full page"]')
let cloneNode = sourceNode.cloneNode(true)           // clone the original button
sourceNode.parentNode.appendChild(cloneNode)         // and insert next to the original button
sourceNode.style.display = "none"                    // while hiding the original button
cloneNode.addEventListener("click", () => {
	if(!document.fullscreenElement) {
		panZoomRef.requestFullscreen()
	} else {
		document.exitFullscreen()
	}
})
```

This will break any hover states I believe, but for our use case we don't use hover states.

It would be nice if this behaviour was more obvious, or if an additional simple mode doing what we do above was available.

**Comment by iangilman** [comment]
@R4wizard Thank you for sharing this! I actually think this is a pretty good approach.

Someone still needs to clean up our fullscreen in OSD (basically to simplify it), as you say.
